doc/manpages/buildcheck.1
.TH dose3-buildcheck 1 "December 4, 2010" "Version 2.9" "USER COMMANDS"
.SH NAME
dose3-buildcheck \- check installability of build-dependencies

.SH SYNOPSIS
\fBdose3-buildcheck\fR [option] ... \fIbinary-repository\fR \fIsource-repository\fR

.SH DESCRIPTION
.B dose3-buildcheck
determines, for a set of debian source package control stanzas, called
the source repository, whether a build environment for the packages of
the source repository can be installed by using packages from the
binary repository. For this, only package meta-information is taken
into account: build-dependencies and build-conflicts in the source
package, and inter-package relationsships expressed in the binary
repository.
.P
The constraint solving algorithm is complete, that is it finds a
solution whenever there exists one, even for multiple disjunctive
dependencies and deep package conflicts. This problem is
computationally infeasible in theory (that is, NP-complete), but can
be solved very efficiently for package repositories that actually occur
in practice.
.P
Installability of binary packages is analyzed according to their
\fBDepends\fR, \fBConflicts\fR, and \fBProvides\fR fields with their
meaning as of Debian policy version 3.9.0. \fBPre-depends\fR are
treated like \fBDepends\fR, and \fBBreaks\fR are treated like
\fBConflicts\fR.

.SH INPUT FORMAT

The \fBbinary-repository\fR argument is the name of a file containing
stanzas in the format of deb-control(5), separated by one blank
line. For instance, the Packages files as found on a Debian mirror
server, or in the directory \fI/var/lib/apt/lists/\fR of a Debian
system, are suitable.
.P
The \fBsource-repository\fR argument is the name of a file containing
debian source control stanzas, separated by one blank
line. For instance, the Sources files as found on a Debian mirror
server, or in the directory \fI/var/lib/apt/lists/\fR of a Debian
system, are suitable.

.SH OPTIONS
.TP
.B \-a \-\-architecture 
Specify the architecture for which installability of build
environments should be checked. This argument is (in the current
version of this program) mandatory.
.TP
.B \-f \-\-failures
Only show broken packages that fail the installability check.
.TP
.B \-v \-\-successes
Only show packages that do not fail the installability check.
.TP
.B \-e \-\-explain
Explain the results in more detail.
.TP
.B \-v
Enable info / warnings / debug messages. This option may be repeated
up to three times in order to increase verbosity.
.TP
.B \-h, \-\-help
Display this list of options.

.SH EXAMPLE
Compute the list of source packages in Sources for which it is not possible
to install a build environment on i386, assuming that the binary packages
described in file Packages are available:

dose3-builddebcheck --architecture i386 --failures --explain Packages Sources

.SH AUTHOR
The current version has been rewritten on the basis of the dose3
library by Pietro Abate; it replaces an earlier version that was 
simply a wrapper for edos-distcheck.

.SH SEE ALSO
.BR deb-control (5),
.BR dose3-distcheck (1)

<http://www.edos-project.org> is the home page of the EDOS project.
<http://www.mancoosi.org> is the home page of the Mancoosi project.



applications/buildcheck.ml
(**************************************************************************************)
(*  Copyright (C) 2009 Pietro Abate <pietro.abate@pps.jussieu.fr>                     *)
(*  Copyright (C) 2009 Mancoosi Project                                               *)
(*                                                                                    *)
(*  This library is free software: you can redistribute it and/or modify              *)
(*  it under the terms of the GNU Lesser General Public License as                    *)
(*  published by the Free Software Foundation, either version 3 of the                *)
(*  License, or (at your option) any later version.  A special linking                *)
(*  exception to the GNU Lesser General Public License applies to this                *)
(*  library, see the COPYING file for more information.                               *)
(**************************************************************************************)

open ExtLib
open Debian
open Common
module Src = Debian.Sources
module Deb = Debian.Packages

module Options = struct
  open OptParse

  let verbose = StdOpt.incr_option ()
  let successes = StdOpt.store_true ()
  let failures = StdOpt.store_true ()
  let explain = StdOpt.store_true ()
  let xml = StdOpt.str_option ()
  let architecture = StdOpt.str_option ()

  let showall () = (Opt.get successes) && (Opt.get failures)
  let onlyfail () = (Opt.get failures) && not (Opt.get successes)
  let onlysucc () = (Opt.get successes) && not (Opt.get failures)

  let description = "Report the broken packages in a package list"
  let options = OptParser.make ~description ()

  open OptParser
  add options ~short_name:'v' ~help:"Print information (can be repeated)" verbose;
  add options ~short_name:'e' ~long_name:"explain" ~help:"Explain the results" explain;
  add options ~short_name:'f' ~long_name:"failures" ~help:"Only show failures" failures;
  add options ~short_name:'s' ~long_name:"successes" ~help:"Only show successes" successes;
  add options ~short_name:'a' ~long_name:"architecture" ~help:"" architecture;
  add options ~long_name:"xml" ~help:"Output results in XML format" xml;
end

let debug fmt = Util.make_debug "Buildcheck" fmt
let info fmt = Util.make_info "Buildcheck" fmt
let warning fmt = Util.make_warning "Buildcheck" fmt

open Diagnostic

let main () =
  let posargs = OptParse.OptParser.parse_argv Options.options in
  Boilerplate.enable_debug (OptParse.Opt.get Options.verbose);

  if not(OptParse.Opt.is_set Options.architecture) then begin
    Printf.eprintf "--architecture must be specified\n";
    exit 1;
  end ;

  let pkglist = Deb.input_raw [List.hd posargs] in
  let srclist = 
    let l = Src.input_raw (List.tl posargs) in
    Src.sources2packages (OptParse.Opt.get Options.architecture) l 
  in
  let tables = Debcudf.init_tables (srclist @ pkglist) in
 
  let sl = List.map (fun pkg -> Debcudf.tocudf tables pkg) srclist in
  let l = List.fold_left (fun acc pkg -> (Debcudf.tocudf tables pkg)::acc) sl pkglist in
  let universe = Cudf.load_universe l in
  let pp pkg =
    let (p,i) = (pkg.Cudf.package,pkg.Cudf.version) in
    let v = Debian.Debcudf.get_real_version tables (p,i) in
    let l =
      List.filter_map (fun k ->
        try Some(k,Cudf.lookup_package_property pkg k)
        with Not_found -> None
      ) ["architecture";"source";"sourceversion"]
    in (p,v,l)
  in
  info "Solving..." ;
  let failure = OptParse.Opt.get Options.failures in
  let success = OptParse.Opt.get Options.successes in
  let explain = OptParse.Opt.get Options.explain in
  let callback = Diagnostic.printf ~pp ~failure ~success ~explain in
  let i = Depsolver.listcheck ~callback universe sl in
  Printf.eprintf "Broken Packages: %d\n" i
;;

main () ;;
